openjdk/barriers/mmtkSATBBarrier.cpp

```diff
@@ -0,0 +1,536 @@
+#endif
+};
+
+void MMTkSATBBarrierSetRuntime::object_probable_write(oop new_obj) const {
+ // The slow-call will do the unlog bit check again (same as the above fast-path check)
+  mmtk_object_probable_write((MMTk_Mutator) &Thread::current()->third_party_heap_mutator, (void*) new_obj);
+}
+
+void MMTkSATBBarrierSetRuntime::object_reference_write_pre(oop src, oop* slot, oop target) const {
+#if MMTK_ENABLE_BARRIER_FASTPATH
+  // oop pre_val = *slot;
+  // if (pre_val == NULL) return;
+  intptr_t addr = ((intptr_t) (void*) src);
+  const volatile uint8_t * meta_addr = (const volatile uint8_t *) (side_metadata_base_address() + (addr >> (UseCompressedOops ? 5 : 6)));
+  intptr_t shift = (addr >> (UseCompressedOops ? 2 : 3)) & 0b111;
+  uint8_t byte_val = *meta_addr;
+  if (((byte_val >> shift) & 1) == kUnloggedValue) {
+    object_reference_write_slow_call((void*) src, (void*) slot, (void*) target);
+  }
+#else
+  object_reference_write_pre_call((void*) src, (void*) slot, (void*) target);
+#endif
+}
+
+#define __ masm->
+
+void MMTkSATBBarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet decorators, BasicType type, Register dst, Address src, Register tmp1, Register tmp_thread) {
+  bool on_oop = type == T_OBJECT || type == T_ARRAY;
+  bool on_weak = (decorators & ON_WEAK_OOP_REF) != 0;
+  bool on_phantom = (decorators & ON_PHANTOM_OOP_REF) != 0;
+  bool on_reference = on_weak || on_phantom;
+  BarrierSetAssembler::load_at(masm, decorators, type, dst, src, tmp1, tmp_thread);
+#if SOFT_REFERENCE_LOAD_BARRIER
+  if (on_oop && on_reference) {
+    Label done;
+    // No slow-call if SATB is not active
+    Register tmp = rscratch1;
+    Register tmp2 = rscratch2;
+    __ movptr(tmp, intptr_t(&CONCURRENT_MARKING_ACTIVE));
+    __ xorq(tmp2, tmp2);
+    __ movb(tmp2, Address(tmp, 0));
+    __ cmpptr(tmp2, 1);
+    __ jcc(Assembler::notEqual, done);
+    // No slow-call if dst is NULL
+    __ cmpptr(dst, 0);
+    __ jcc(Assembler::equal, done);
+    // Do slow-call
+    __ pusha();
+    __ mov(c_rarg0, dst);
+    __ MacroAssembler::call_VM_leaf_base(FN_ADDR(MMTkBarrierSetRuntime::load_reference_call), 1);
+    __ popa();
+    __ bind(done);
+  }
+#endif
+}
+
+void MMTkSATBBarrierSetAssembler::object_reference_write_pre(MacroAssembler* masm, DecoratorSet decorators, Address dst, Register val, Register tmp1, Register tmp2) const {
+  if (can_remove_barrier(decorators, val, /* skip_const_null */ false)) return;
+
+  #if MMTK_ENABLE_BARRIER_FASTPATH
+  Label done;
+
+  Register obj = dst.base();
+  Register tmp3 = rscratch1;
+  Register tmp4 = rscratch2;
+  Register tmp5 = tmp1 == dst.base() || tmp1 == dst.index() ? tmp2 : tmp1;
+
+  // tmp5 = load-byte (side_metadata_base_address() + (obj >> 6));
+  __ movptr(tmp3, obj);
+  // __ load_heap_oop(tmp3, dst, noreg, noreg, AS_RAW);
+  // // Is the previous value null?
+  // __ cmpptr(tmp3, (int32_t) NULL_WORD);
+  // __ jcc(Assembler::equal, done);
+
+  __ shrptr(tmp3, UseCompressedOops ? 5 : 6);
```

Comment:
For the same reason, we should just shift by 6 on 64-bit machines.

```suggestion
  __ shrptr(tmp3, 6);
```